File: python/day-2/b.py
def is_valid(line):
    space = line.find(' ')
    dash = line.find('-')

    nums = line[:space]
    lower = int(nums[:dash]) - 1
    upper = int(nums[dash + 1:]) - 1
    letter = line[space + 1: space + 2]
    password = line[line.find(':') + 2:]

    return (password[lower] == letter) ^ (password[upper] == letter)
        
data = open('input.txt').read().split('\n')[:-1]
print(sum(is_valid(line) for line in data))

File: rust/day-2/b.rs
use std::fs;
use std::error::Error;

fn main() -> Result<(), Box<dyn Error>> {
    let inp = fs::read_to_string("input.txt")?;

    let count = inp
        .lines()
        .filter(|line| !line.is_empty() && is_valid(line))
        .count();
    
    println!("{}", count);
    Ok(())

}

fn is_valid(password: &str) -> bool {
    let space = password.find(' ').unwrap();
    let dash = password.find('-').unwrap();
    let colon = password.find(':').unwrap();

    let nums = &password[..space];
    let lower = nums[..dash].parse::<usize>().unwrap() - 1;
    let upper = nums[dash+1..].parse::<usize>().unwrap() - 1;

    let letter = password.chars().nth(space + 1).unwrap();
    let pass = &password[colon+2..];

    let c1 = pass.chars().nth(lower).unwrap();
    let c2 = pass.chars().nth(upper).unwrap();

    (c1 == letter) ^ (c2 == letter)
}